hub3/ead/dao.go

```diff
@@ -0,0 +1,622 @@
+}
+
+func (cfg *DaoConfig) getConfigPath() string {
+	return getDaoConfigPath(cfg.ArchiveID, cfg.UUID)
+}
+
+func (cfg *DaoConfig) getDirPath() string {
+	return getMetsDirPath(cfg.ArchiveID)
+}
+
+func (cfg *DaoConfig) Write() error {
+	err := os.MkdirAll(cfg.getDirPath(), os.ModePerm)
+	if err != nil {
+		log.Error().Err(err).Msg("mkdir error")
+		return err
+	}
+
+	b, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(
+		cfg.getConfigPath(),
+		b,
+		os.ModePerm,
+	)
+}
+
+// Delete removes the DaoConfig and METS file
+func (cfg *DaoConfig) Delete() error {
+	files := []string{
+		cfg.getMetsFilePath(),
+		getDaoConfigPath(cfg.ArchiveID, cfg.UUID),
+	}
+
+	for _, f := range files {
+		if err := os.Remove(f); err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cfg *DaoConfig) hasOrphanedMetsFile() bool {
+	if cfg.metsFileExists() && !cfg.daoConfigExists() {
+		return true
+	}
+
+	return false
+}
+
+func (cfg *DaoConfig) daoConfigExists() bool {
+	_, err := os.Stat(cfg.getConfigPath())
+
+	return !os.IsNotExist(err)
+}
+
+func (cfg *DaoConfig) metsFileExists() bool {
+	_, err := os.Stat(cfg.getMetsFilePath())
+
+	return !os.IsNotExist(err)
+}
+
+func (cfg *DaoConfig) Mets() (*Cmets, error) {
+	return readMETS(cfg.getMetsFilePath())
+}
+
+func (cfg *DaoConfig) FindingAid(c *DaoClient) (eadpb.FindingAid, error) {
+	if c.HttpFallback && !cfg.metsFileExists() {
+		if err := c.StoreMets(cfg); err != nil {
+			return eadpb.FindingAid{}, err
+		}
+	}
+
+	mets, err := cfg.Mets()
+	if err != nil {
+		return eadpb.FindingAid{}, err
+	}
+
+	return mets.newFindingAid(cfg)
+}
+
+func createHeader(cfg *DaoConfig, id, subjectBase, tag string) *fragments.Header {
+	subject := fmt.Sprintf("%s/%s", subjectBase, id)
+	header := &fragments.Header{
+		OrgID: cfg.OrgID,
+		Spec:  cfg.ArchiveID,
+		HubID: fmt.Sprintf(
+			"%s_%s_%s",
+			cfg.OrgID,
+			cfg.ArchiveID,
+			strings.ReplaceAll(id, "/", "-"),
+		),
+		DocType:       fragments.FragmentGraphDocType,
+		EntryURI:      subject,
+		NamedGraphURI: fmt.Sprintf("%s/graph", subject),
+		SourcePath:    "mets/" + cfg.UUID,
+		GroupID:       cfg.InventoryPath,
+		Modified:      fragments.NowInMillis(),
+		Tags:          []string{tag},
+	}
+
+	return header
+}
+
+func (cfg *DaoConfig) getArchiveTitle() string {
+	if len(cfg.ArchiveTitle) == 0 || cfg.ArchiveTitle[0] == "" {
+		return ""
+	}
+
+	return cfg.ArchiveTitle[0]
+}
+
+func (cfg *DaoConfig) fileTriples(subject string, file *eadpb.File) []*rdf.Triple {
+	s := rdf.NewResource(subject)
+	triples := []*rdf.Triple{
+		rdf.NewTriple(
+			s,
+			rdf.NewResource(fragments.RDFType),
+			rdf.NewResource("https://archief.nl/def/ead/mets/File"),
+		),
+		rdf.NewTriple(
+			s,
+			rdf.NewResource("http://www.w3.org/2000/01/rdf-schema#label"),
+			rdf.NewLiteral(file.Filename),
+		),
+	}
+	t := func(s rdf.Term, p, o string, oType convert) {
```

Comment:
1] How likely is it that the `convert` parameter will contain a non `rdf.NewLiteral` in the near future?
2] I don't mind the function being named `t` since the scope is so small, but if it were named something like `appendTriple` I would have to expand even less mental energy while reading related calls

Reply:
ad 1) It is at least an option that I would not like to exclude.

ad 2) I have used this pattern more often to focus readability what we are adding.  I have made a change in the function that I will push shortly.